[mempool] Mark mempool import fails that were found in mempool as 'already there'

kallewoof opened the pull request:
I was investigating the reasons for failed imports in mempool and noticed that `LoadMempool()` and `pwallet->postInitProcess()` (for all wallets) are executed concurrently. The wallet will end up importing transactions that `LoadMempool()` later tries to import; the latter will fail due to the tx already being in the mempool.

This PR changes the log message, adding an additional "already there" entry. For transactions not accepted into mempool, a check if they are in the mempool is done first, and if found, they are counted as 'already there', otherwise counted as 'failed'.

Also slight rewording for consistency (successes, failed, expired, ... -> succeeded, failed, expired).

kallewoof committed:
[mempool] Mark unaccepted txs present in mempool as 'already there'.

On startup, the wallets will start pumping wallet transactions into the mempool in a different thread while LoadMempool() is running.
This will sometimes result in transactions "failing" to be accepted into mempool, but only for the reason that they were already
put there by a wallet. The log message for mempool load would note this as a 'failure' to import, which was misleading; it should
instead mark it as the transaction already being in the mempool.

promag commented on src/validation.cpp:
```diff
@@ -4303,7 +4304,15 @@ bool LoadMempool(void)
+                    // wallet(s) having loaded it while we were processing
+                    // mempool transactions; consider these as valid, instead of
+                    // failed, but mark them as 'already existed'
+                    if (mempool.exists(tx->GetHash())) {
```
Check `state.GetRejectCode() == REJECT_DUPLICATE`? See [this](https://github.com/bitcoin/bitcoin/blame/master/src/validation.cpp#L476).

Also, IMO this is prettier:
```cpp
if (state.IsValid()) {
    ...
} else if (state.GetRejectCode() == REJECT_DUPLICATE) {
    ...
} else {
    ...
}
```

promag replied:
There is more than one reason for`REJECT_DUPLICATE`, don't know if that matters here. @sipa?

sipa replied:
REJECT_DUPLICATE is used to indicate double spends, so that wouldn't be correct.

promag replied:
What about #10503?

promag replied:
But REJECT_DUPLICATE is set if it's in the mempool which happens to be what's tested here. So it should be?
```cpp
} else if (state.GetRejectCode() == REJECT_DUPLICATE && state.GetRejectReason() == "txn-already-in-mempool") {
```

sipa replied:
That's pretty ugly. There's no good way to distinuish replay from double spend... for almost everything the distinction doesn't matter.

promag replied:
Got it. So in this case `if (mempool.exists(tx->GetHash())` can come before `AcceptToMemoryPoolWithTime`?

kallewoof replied:
As @sipa noted, `REJECT_DUPLICATE` is ambiguous. I don't want to say it was already there, when a conflicted different transaction was there instead. The safest course would seem to simply check if it is there after a failure, as I am doing now.

promag replied:
Understood, can't use REJECT_DUPLICATE.

Still, the way it's now is like catching the error, whereas testing before it's like expecting that behaviour.

kallewoof replied:
Yes. There's a race condition going on here, which makes this better than the alternative (i.e. checking if exists and then trying to accept).

Thread A is doing wallet mempool insertions.
Thread B is calling LoadMempool().

Case 1:
- Thread A imports tx X
- Thread B tests for presence of tx X. Found. Marks already there and moves on.

Case 2:
- Thread B tests for presence of tx X. Not found.
- Thread A imports tx X.
- Thread B attempts to import tx X and fails.

Both cases should result in 'already there', but by checking presence before attempt to accept, the second case will result in a failure.

promag replied:
Doesn't the above `LOCK(cs_main)` prevents that case (and the current implementation for all that matters)?

kallewoof replied:
Hm, you're right, it does. I think.

promag commented on src/validation.cpp:
```diff
@@ -4274,6 +4274,7 @@ bool LoadMempool(void)
     int64_t count = 0;
     int64_t skipped = 0;
     int64_t failed = 0;
+    int64_t alreadythere = 0;
```
Maybe just increment `skipped`?

kallewoof replied:
But it wasn't skipped. It was already there.

promag replied:
Rigth, maybe rename skipped to expired.

kallewoof replied:
Good point. Done.

promag reviewed (commented):
Switch commit order? Commit messages are a bit long, maybe reduce?

sipa commented:
Yes, what about it?

sipa commented:
utACK

TheBlueMatt commented:
utACK 8c8ef83edc8724f2dc7e363e9c412c5df564e08f

MarcoFalke commented:
utACK 8c8ef83

jnewbery commented on src/validation.cpp:
```diff
@@ -4272,8 +4272,9 @@ bool LoadMempool(void)
-    int64_t skipped = 0;
+    int64_t expired = 0;
     int64_t failed = 0;
+    int64_t alreadythere = 0;
```
nit: snake case `already_there`

promag replied:
IMO consider renaming to one of `already_imported`, `already_loaded`, `already_in_mempool`.

kallewoof replied:
Personally feel the current name is fine as is, but will change if people want it.

jnewbery commented on src/validation.cpp:
```diff
@@ -4303,10 +4304,18 @@ bool LoadMempool(void)
+                    // mempool may contain the transaction already, e.g. from
+                    // wallet(s) having loaded it while we were processing
+                    // mempool transactions; consider these as valid, instead of
+                    // failed, but mark them as 'already existed'
```
nit: mark them as 'already there'

jnewbery reviewed (commented):
Seems reasonable. utACK 8c8ef83edc8724f2dc7e363e9c412c5df564e08f , although I think the commits can be squashed down into one commit for merge. I have a couple of supernits on variable names and comments which you can take or leave.

Shame there aren't any LoadMempool unit tests that you could add to. I think adding them from scratch is beyond the scope of this PR.

kallewoof commented:
@jnewbery Thanks for the review! You're right, I think I was overthinking things when I tried to split commits. I've squashed into one commit and addressed your nits.

jnewbery commented:
`already_there` seems fine to me. Will retest later.

kallewoof commented:
Didn't realize test failed. Restarting as it's obviously unrelated.

jnewbery commented:
Tested ACK 258d33b41a27917f59e3aee856d032a23cbb5b05

MarcoFalke commented:
re-utACK 258d33b

promag commented:
utACK 258d33b.
